test/Microsoft.Azure.SignalR.Tests/Infrastructure/TestHub.cs

```diff
@@ -1,15 +1,88 @@
 
 namespace Microsoft.Azure.SignalR.Tests
 {
-    public class TestHub : Hub
+    internal class TestHub : Hub
     {
+        private static ConcurrentDictionary<string, bool> _connectedConnections = new ConcurrentDictionary<string, bool>();
+        private static ConcurrentDictionary<string, bool> _connectedUsers = new ConcurrentDictionary<string, bool>();
+
+        public static void ClearConnectedConnectionAndUser()
+        {
+            _connectedConnections.Clear();
+            _connectedUsers.Clear();
+        }
+
+        public override Task OnConnectedAsync()
+        {
+            if (!_connectedConnections.TryAdd(Context.ConnectionId, false))
+            {
+                throw new InvalidOperationException($"Failed to add a client connection {Context.ConnectionId}. Connected connections {string.Join(",", _connectedConnections.Keys)}.");
+            }
+
+            if (!_connectedUsers.TryAdd(Context.UserIdentifier, false))
+            {
+                throw new InvalidOperationException($"Failed to add connection {Context.ConnectionId} as user {Context.UserIdentifier}. Connected users: {string.Join(", ", _connectedUsers.Keys)}");
+            }
+            return Task.CompletedTask;
+        }
+
+        public override Task OnDisconnectedAsync(Exception exception)
+        {
+            if (!_connectedConnections.TryRemove(Context.ConnectionId, out _))
+            {
+                throw new InvalidOperationException($"Failed to remove a client connection {Context.ConnectionId}. Connected connections {string.Join(",", _connectedConnections.Keys)}.");
+            }
+
+            if (!_connectedUsers.TryRemove(Context.UserIdentifier, out _))
+            {
+                throw new InvalidOperationException($"Failed to remove a client connection {Context.ConnectionId} for as user {Context.UserIdentifier}. Connected users: {string.Join(", ", _connectedUsers.Keys)}");
+            }
+            return Task.CompletedTask;
+        }
+
         public void Echo(string message)
         {
-            Clients.Client(Context.ConnectionId).SendAsync(nameof(Echo), message);
+            Clients.Caller.SendAsync(nameof(Echo), message);
+        }
+
+        public void SendToClient(string message)
+        {
+            var ind = StaticRandom.Next(0, _connectedConnections.Count);
+            Clients.Client(_connectedConnections.Keys.ToList()[ind]).SendAsync(nameof(SendToClient), message);
+        }
+
+        public void SendToUser(string message)
+        {
+            var ind = StaticRandom.Next(0, _connectedUsers.Count);
+            Clients.User(_connectedUsers.Keys.ToList()[ind]).SendAsync(nameof(SendToUser), message);
+        }
+
+        public void Broadcast(string message)
+        {
+            Clients.All.SendAsync(nameof(Broadcast), message);
+        }
+
+        public void JoinGroup(string groupName)
+        {
+            Groups.AddToGroupAsync(Context.ConnectionId, groupName);
```

Comment:
`AddToGroupAsync` and `RemoveFromGroupAsync` can not guarantee order now if they're not `await`. Do we need to change to `public Task JoinGroup()`

Reply:
But `await Groups.AddToGroupAsync(Context.ConnectionId, groupName);` still cannot guarantee the order, later messages might be sent in other server connections, is that right?

For now, I await `DefaultDelayMilliseconds` seconds after join/leave group to "make sure" the connection joined/leaved the group.

Reply:
I've merged ack-able group, so it can guarantee the order now.

Reply:
great!